django/tasks/task.py

```diff
@@ -0,0 +1,271 @@
+from dataclasses import dataclass, field, replace
+from datetime import datetime
+from inspect import iscoroutinefunction
+from typing import Any, Callable, Dict, Optional, Type
+
+from asgiref.sync import async_to_sync, sync_to_async
+
+from django.db.models.enums import TextChoices
+from django.utils.translation import gettext_lazy as _
+
+from .exceptions import ResultDoesNotExist
+from .utils import get_module_path, json_normalize
+
+DEFAULT_TASK_BACKEND_ALIAS = "default"
+DEFAULT_QUEUE_NAME = "default"
+MIN_PRIORITY = -100
+MAX_PRIORITY = 100
+DEFAULT_PRIORITY = 0
+
+TASK_REFRESH_ATTRS = {
+    "_exception_class",
+    "_traceback",
+    "_return_value",
+    "finished_at",
+    "started_at",
+    "status",
+    "enqueued_at",
+}
+
+
+class ResultStatus(TextChoices):
+    NEW = ("NEW", _("New"))
+    RUNNING = ("RUNNING", _("Running"))
+    FAILED = ("FAILED", _("Failed"))
+    SUCCEEDED = ("SUCCEEDED", _("Succeeded"))
+
+
+@dataclass(frozen=True)
+class Task:
+    priority: int
+    """The priority of the task"""
+
+    func: Callable
+    """The task function"""
+
+    backend: str
+    """The name of the backend the task will run on"""
+
+    queue_name: str = DEFAULT_QUEUE_NAME
+    """The name of the queue the task will run on"""
+
+    run_after: Optional[datetime] = None
+    """The earliest this task will run"""
+
+    enqueue_on_commit: Optional[bool] = None
+    """
+    Whether the task will be enqueued when the current transaction commits,
+    immediately, or whatever the backend decides
+    """
+
+    def __post_init__(self):
+        self.get_backend().validate_task(self)
+
+    @property
+    def name(self):
+        """
+        An identifier for the task
+        """
+        return self.func.__name__
+
+    def using(
+        self,
+        *,
+        priority=None,
+        queue_name=None,
+        run_after=None,
+        backend=None,
+    ):
+        """
+        Create a new task with modified defaults
+        """
+
+        changes = {}
+
+        if priority is not None:
+            changes["priority"] = priority
+        if queue_name is not None:
+            changes["queue_name"] = queue_name
+        if run_after is not None:
+            changes["run_after"] = run_after
+        if backend is not None:
+            changes["backend"] = backend
+
+        return replace(self, **changes)
+
+    def enqueue(self, *args, **kwargs):
+        """
+        Queue up the task to be executed
+        """
+        return self.get_backend().enqueue(
+            self, json_normalize(args), json_normalize(kwargs)
```

Comment:
Maybe the `json_normalize` can be done only when `DEBUG` or something like that? It's a bit unfortunate...

Reply:
`json_normalize` is still necessary, but  I agree in this case it can be handled by the backend rather than with each task.

It's mostly there to ensure that the `.args` and `.kwargs` attributes are set correctly.